workflow/scripts/create_transport_network.py

```diff
@@ -0,0 +1,247 @@
+
+
+def cast(x: Any, *, casting_function: Callable, nullable: bool) -> Any:
+    """
+    Attempt to recast value with provided function. If not possible and
+    nullable is true, return None. Else, raise casting error.
+
+    N.B. Empty string is not considered a nullable value.
+
+    Args:
+        x (Any): Value to cast
+        casting_function (Callable): Function to cast with
+        nullable (bool): Whether cast value can be None
+
+    Returns:
+        Any: Recast value
+    """
+
+    try:
+        new_value = casting_function(x)
+
+        if new_value is None and not nullable:
+            raise TypeError(f"{new_value=} with {casting_function=} and {nullable=}")
+        else:
+            return new_value
+
+    except (ValueError, TypeError) as casting_error:
+        # ValueError in case of e.g. x="50 mph"
+        # TypeError in case of e.g. x=None
+
+        if nullable:
+            return None
+        else:
+            raise ValueError("Couldn't recast to non-nullable value") from casting_error
+
+
+def clean_OSM_ways(ways: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+    """
+    Check and clean OpenStreetMap input data
+
+    Args:
+        ways (gpd.GeoDataFrame): Table of ways created by osmium and osm_to_pq.py
+
+    Returns:
+        gpd.GeoDataFrame: Cleaned table
+    """
+
+    # check we have exactly the columns we expect
+    necessary_columns = {
+        'geometry',
+        'tag_highway',
+        'tag_surface',
+        'tag_bridge',
+        'tag_maxspeed',
+        'tag_lanes',
+    }
+    if not necessary_columns.issubset(set(ways.columns.values)):
+        raise ValueError(f"{ways.columns.values} does not contain {set(ways.columns.values) - necessary_columns}")
+
+    # check we have only linestrings
+    assert set(ways.geometry.type.values) == {'LineString'}
+
+    # recast data where appropriate
+    # be careful applying this to string fields, your no-data values may change
+    type_conversion_data = (
+        ('tag_maxspeed', float, True),
+        ('tag_lanes', int, True),
+    )
+    for column_name, dtype, nullable in type_conversion_data:
+        ways[column_name] = ways[column_name].apply(cast, casting_function=dtype, nullable=nullable)
+
+    # make the bridge tag boolean (on read the values are almost entirely None or 'yes')
+    ways.tag_bridge = ways.tag_bridge.apply(bool)
+
+    # turn the <highway_type>_link entries into <highway_type>
+    ways.tag_highway = ways.tag_highway.apply(strip_suffix)
+
+    # drop the 'tag_' prefix added during osm.pbf processing
+    # first check there won't be a collision with new column names
+    ways_renamed = ways.rename(strip_prefix, axis='columns')
+    n_cols_renamed: int = len(set(ways_renamed.columns.values))
+    if n_cols_renamed == len(set(ways.columns.values)):
+        ways = ways_renamed
+    else:
+        raise ValueError(f"Removing prefix would result in collision: {ways.columns=}")
+
+    return ways
+
+
+def create_network(
```

Comment:
Do you think `create_network` should be pulled out to `snkit` as a sort of general fix-me-up top-level function? It seems like we often end up with roughly this series of calls to snkit functions.

Reply:
I quite like explicitly including each step for maximum visibility of which cleaning processes you're employing.